### PR TITLE
BUG make hashtable.unique support readonly arrays

### LINKED ISSUE #18773: pd.cut fails on readonly arrays
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import numpy as np

bins = np.arange(0, 100, 10)
bins.flags.writeable = False
pd.cut(np.arange(100), bins)
```
Results in:
```
ValueError                                Traceback (most recent call last)
<ipython-input-71-60b85be1194d> in <module>()
----> 1 pd.cut(np.arange(100), bins)

~/conda/envs/xarray-py36/lib/python3.6/site-packages/pandas/core/reshape/tile.py in cut(x, bins, right, labels, retbins, precision, include_lowest)
    134                               precision=precision,
    135                               include_lowest=include_lowest,
--> 136                               dtype=dtype)
    137
    138     return _postprocess_for_cut(fac, bins, retbins, x_is_series,

~/conda/envs/xarray-py36/lib/python3.6/site-packages/pandas/core/reshape/tile.py in _bins_to_cuts(x, bins, right, labels, precision, include_lowest, dtype, duplicates)
    225         return result, bins
    226
--> 227     unique_bins = algos.unique(bins)
    228     if len(unique_bins) < len(bins) and len(bins) != 2:
    229         if duplicates == 'raise':

~/conda/envs/xarray-py36/lib/python3.6/site-packages/pandas/core/algorithms.py in unique(values)
    354
    355     table = htable(len(values))
--> 356     uniques = table.unique(values)
    357     uniques = _reconstruct_data(uniques, dtype, original)
    358

pandas/_libs/hashtable_class_helper.pxi in pandas._libs.hashtable.Int64HashTable.unique (pandas/_libs/hashtable.c:16341)()

~/conda/envs/xarray-py36/lib/python3.6/site-packages/pandas/_libs/hashtable.cpython-36m-darwin.so in View.MemoryView.memoryview_cwrapper (pandas/_libs/hashtable.c:45205)()

~/conda/envs/xarray-py36/lib/python3.6/site-packages/pandas/_libs/hashtable.cpython-36m-darwin.so in View.MemoryView.memoryview.__cinit__ (pandas/_libs/hashtable.c:41440)()

ValueError: buffer source array is read-only
```

#### Problem description

This is essentially the same problem as https://github.com/pandas-dev/pandas/issues/10043 and https://github.com/pandas-dev/pandas/issues/17192, due to Cython issue https://github.com/cython/cython/issues/1605

#### Expected Output

Should not error.

#### Output of ``pd.show_versions()``

<details>

In [73]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Darwin
OS-release: 16.7.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.20.3
pytest: 3.2.1
pip: 9.0.1
setuptools: 36.4.0
Cython: 0.26
numpy: 1.13.1
scipy: 0.19.1
xarray: 0.10.0
IPython: 6.1.0
sphinx: 1.6.3
patsy: None
dateutil: 2.6.1
pytz: 2017.2
blosc: None
bottleneck: 1.2.1
tables: None
numexpr: None
feather: None
matplotlib: 2.0.2
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.9.6
s3fs: None
pandas_gbq: None
pandas_datareader: None
</details>

### PR DESCRIPTION
**Pull request by hexgnu**
This problem was brought up in
https://github.com/pandas-dev/pandas/issues/18773 and effectively comes
down to how Cython deals with readonly arrays. While it would be ideal
for Cython to fix the underlying problem in the meantime we can rely on
this.

- [x] closes #18773
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/reshape/test_tile.py [hunk 1]
```diff
@@ -512,6 +512,16 @@ def f():
         tm.assert_numpy_array_equal(
             mask, np.array([False, True, True, True, True]))
 
+    def test_cut_read_only(self):
```
Comment (gfyoung) [inline pandas/tests/reshape/test_tile.py]: Reference issue number below.

Comment (hexgnu) [reply pandas/tests/reshape/test_tile.py]: done thanks @gfyoung

File: pandas/tests/reshape/test_tile.py [hunk 2]
```diff
@@ -512,6 +512,17 @@ def f():
...
+        # issue 18773
+        readonly = np.arange(0, 100, 10)
+        readonly.flags.writeable = False
+
+        mutable = np.arange(0, 100, 10)
+
+        one_to_hundred = np.arange(100)
```
Comment (gfyoung) [inline pandas/tests/reshape/test_tile.py]: Not quite...that's really zero to 99 :smile:

Comment (hexgnu) [reply pandas/tests/reshape/test_tile.py]: oh duh lol... my bad I'll update it

File: pandas/tests/reshape/test_tile.py [hunk 3]
```diff
@@ -512,6 +512,17 @@ def f():
...
 
+    def test_cut_read_only(self):
+        # issue 18773
+        readonly = np.arange(0, 100, 10)
+        readonly.flags.writeable = False
+
+        mutable = np.arange(0, 100, 10)
+
+        hundred_elements = np.arange(100)
+        tm.assert_categorical_equal(cut(hundred_elements, readonly),
```
Comment (jreback) [inline pandas/tests/reshape/test_tile.py]: you can paramaterize on writeable here.

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 4]
```diff
@@ -255,10 +255,56 @@ dtypes = [('Float64', 'float64', 'val != val', True),
           ('UInt64', 'uint64', 'False', False),
           ('Int64', 'int64', 'val == iNaT', False)]
 
+def get_dispatch(dtypes):
```
Comment (jreback) [inline pandas/_libs/hashtable_class_helper.pxi.in]: can you add a comment similiar to in algos_take for using the memory view.

**Review by jreback — CHANGES_REQUESTED** [review]
small comments. Ideally we would find a better / more general way of doing this.

File: pandas/_libs/hashtable_class_helper.pxi.in (hunk 4 above)
Comment (hexgnu) [reply pandas/_libs/hashtable_class_helper.pxi.in]: Done. I also noticed that the indenting was a tad off so I re-indented it to be 4 spaces.

File: pandas/tests/reshape/test_tile.py (hunk 3 above)
Comment (hexgnu) [reply pandas/tests/reshape/test_tile.py]: I think this is what you're asking. I added a parameterized test that tests the three combinations (Readonly, Mutable) (Mutable, Mutable) (Readonly, Readonly). Let me know if that's not what you were thinking. Thanks @jreback

File: pandas/tests/reshape/test_tile.py [hunk 5]
```diff
@@ -512,6 +512,23 @@ def f():
...
+
+            array_2 = np.arange(0, 100, 10)
+            array_2.flags.writeable = array_2_writeable
+
+            hundred_elements = np.arange(100)
+
+            tm.assert_categorical_equal(cut(hundred_elements, array_1),
+                                        cut(hundred_elements, array_2))
+
+        for (w_1, w_2) in [[True, True], [True, False], [False, False]]:
```
Comment (jreback) [inline pandas/tests/reshape/test_tile.py]: can you parametrize this test instead of writing a sub test function

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/reshape/test_tile.py (hunk 5 above)
Comment (hexgnu) [reply pandas/tests/reshape/test_tile.py]: sorry yea just pushed a commit for that. should be parametrized now :smile:

File: doc/source/whatsnew/v0.22.0.txt [hunk 6]
```diff
@@ -326,7 +326,7 @@ Reshaping
 - Bug in :func:`DataFrame.stack` which fails trying to sort mixed type levels under Python 3 (:issue:`18310`)
 - Fixed construction of a :class:`Series` from a ``dict`` containing ``NaN`` as key (:issue:`18480`)
 - Bug in :func:`Series.rank` where ``Series`` containing ``NaT`` modifies the ``Series`` inplace (:issue:`18521`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.22.0.txt]: rebase on master. can you move to 0.23 (docs were renamed), prob easiest to just check this file from master and past in new one

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.22.0.txt (hunk 6 above)
Comment (hexgnu) [reply doc/source/whatsnew/v0.22.0.txt]: Done. I also squashed some of my commits to clean up the commit history for you.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
I pushed a fix. ping on green.

**Comment by hexgnu** [comment]
Awesome thanks @jreback looks like it all passed.

**Comment by jreback** [comment]
thanks @hexgnu 
keep em coming!

**Pull request merged**